[EuiSelectable] Focus option after search value change

myasonik opened the pull request:
### Summary

Fixes [bug report](https://github.com/elastic/eui/pull/3169#pullrequestreview-438550377) from @cchaos:
> The only thing I felt odd in the functionality was when the list was filtered down from the search box, and an option is selected it loses focus where as it would normally stay focused.

Also fixes [follow-up bug](https://github.com/elastic/eui/pull/3670#issuecomment-651872086) also found by @cchaos of selecting an option after searching would lose focus.

### Breaking changes
- `options` passed into **EuiSelectable** cannot have an `id` (an `id` is generated internally)
	- This feels kind of risky but this wasn't documented as a possibility before so it shouldn't be too bad (out of 23 uses in Kibana, none passed in an `id`)
- Requires an `onChange` to be passed into **EuiSelectableSearch**
	- Not meant to be used outside of EUI and isn't documented individually though is exported so probably won't have much impact
	- The component doesn't do anything without it, so simplified the types to expected usage

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3670/

cchaos commented:
Sorry I wasn't super clear in my bug report. The click handles it just fine, but if I'm using the keyboard and use <kbd>enter</kbd> to select/deselect the option, the focus is removed from that item.  In fact, I just started testing this with VO on and I'm really confused about what is **supposed** to happen.

It doesn't read out the options as I arrow through them. It only sometimes(?) reads selected options and only ever 1 of them. Why the text read out of "completion" and not "selected"?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3670/

chandlerprall commented:
<kbd>enter</kbd> works as expected, but now it appears clicking on an option moves focus to the first selected option.

myasonik commented:
@cchaos 
The VoiceOver experience is a definitely a mixed bag right now. I want to keep that discussion out of this PR and focus on just these two bugs. 

I added some comments about the screen reader experience to the [feature branch PR](https://github.com/elastic/eui/pull/3169#issuecomment-652078286). 

@chandlerprall 
Sorry about the low hanging bug! Fixed (I hope)!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3670/

chandlerprall commented:
Mouse+focus interaction is still weird. If the selectable doesn't have focus, mousedown moves focus to the first selected item, and only after mouseup/click is the focus moved to the freshly selected option.

![mouse focus](https://chandler-public.s3-us-west-1.amazonaws.com/focus_flash.gif)

myasonik commented:
Third time's the charm 🤞

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3670/

cchaos commented:
I'm still getting the issue where using the keyboard to select an item on a filtered list, removes the focus from that item.

myasonik commented:
🤔 I'm having trouble reproducing again. I did find a case where it would set index to the wrong option though so I fixed that. 

@cchaos Can you see if that fixed the issue for you? If not, could you take a screencast of the problem?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3670/

cchaos commented:
Unfortunately it does not. Steps to reproduce:

1. Type `d` in the search
2. Arrow down to "Dione"
3. <kbd>enter</kbd> to select the option
4. Notice that the focus is not gone from that option and you have to arrow down to it again.

![Screen Recording 2020-07-02 at 03 38 PM](https://user-images.githubusercontent.com/549577/86402758-5c5f4180-bc7a-11ea-8432-bb98e4edd1c9.gif)

myasonik commented:
Thanks for the screencast @cchaos! Got it fixed!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_3670/

chandlerprall commented:
One last thing: pressing any key that isn't <kbd>←</kbd> or <kbd>→</kbd> shifts focus to the first select option. This is better than before this PR, which shifted focus to an unknown element on keypress. The correct solution should be to match native selection/option boxes and skip to the entry matching typed characters, but that is outside the scope of this PR. If it's easy, can we make keypresses do nothing? Otherwise, these changes look good.

myasonik commented:
I'm not really sure if that is the answer. We have our own filtering setup for the results when you type so trying to preempt the results of the filter by jumping to the first result which starts with the letter typed seems like competing ranking systems.

I tried looking through the spec for any "official" guidance one way or the other and couldn't find anything concrete. 

I got to the current implementation of focusing on the first selected item if available and first item in the results if nothing is selected from the [multi-select listbox keyboard interactions](https://www.w3.org/TR/wai-aria-practices/#listbox_kbd_interaction) description. So though this is in a combobox now which could change what best practices are, I think this is a fine thing to ship with.

If we ever get to the point of doing real user testing, this would be a good candidate to test out.

myasonik commented:
Got the OK from @chandlerprall to merge over slack

<img width="276" alt="Screenshot of Slack where Chandler Prall says 'Sounds good to me! If it's a punt, feel free to merge'" src="https://user-images.githubusercontent.com/4188087/86615017-7b5f1b80-bf79-11ea-833c-800febafc77a.png">
